compiler/src/vm/mod.rs

```diff
@@ -1,8 +1,854 @@
+                        );
+
+                        self.fibers
+                            .get_mut(&parent_id)
+                            .unwrap()
+                            .as_parallel_mut()
+                            .unwrap()
+                            .children
+                            .insert(child_id, ChildKind::SpawnedChild(return_channel));
+                    }
+                    None => self.finish_parallel(
+                        parent_id,
+                        Err("a nursery received an invalid message".to_string()),
+                    ),
+                }
+
+                InternalChannel::complete_send(&mut self.fibers, performing_fiber);
+            }
+        }
+    }
+    fn parse_spawn_packet(packet: Packet) -> Option<(Heap, Pointer, ChannelId)> {
+        let Packet { mut heap, value } = packet;
+        let arguments: Struct = heap.get(value).data.clone().try_into().ok()?;
+
+        let closure_symbol = heap.create_symbol("Closure".to_string());
+        let closure_address = arguments.get(&heap, closure_symbol)?;
+        let closure: Closure = heap.get(closure_address).data.clone().try_into().ok()?;
+        if closure.num_args > 0 {
+            return None;
+        }
+
+        let return_channel_symbol = heap.create_symbol("ReturnChannel".to_string());
+        let return_channel_address = arguments.get(&heap, return_channel_symbol)?;
+        let return_channel: SendPort = heap
+            .get(return_channel_address)
+            .data
+            .clone()
+            .try_into()
+            .ok()?;
+
+        Some((heap, closure_address, return_channel.channel))
+    }
+
+    fn receive_from_channel(&mut self, performing_fiber: Option<FiberId>, channel: ChannelId) {
+        match self.channels.get_mut(&channel).unwrap() {
+            Channel::Internal(channel) => {
+                channel.receive(&mut self.fibers, performing_fiber);
+            }
+            Channel::External => {
+                self.push_external_operation(channel, Operation::Receive { performing_fiber });
+            }
+            Channel::Nursery { .. } => unreachable!("nurseries are only sent stuff"),
+        }
+    }
+
+    fn push_external_operation(&mut self, channel: ChannelId, operation: Operation) {
+        self.external_operations
+            .entry(channel)
+            .or_default()
+            .push(operation);
+    }
+}
+
+impl InternalChannel {
+    fn send(
+        &mut self,
+        fibers: &mut HashMap<FiberId, FiberTree>,
+        performing_fiber: Option<FiberId>,
+        packet: Packet,
+    ) {
+        self.pending_sends.push_back((performing_fiber, packet));
+        self.work_on_pending_operations(fibers);
+    }
+
+    fn receive(
+        &mut self,
+        fibers: &mut HashMap<FiberId, FiberTree>,
+        performing_fiber: Option<FiberId>,
+    ) {
+        self.pending_receives.push_back(performing_fiber);
+        self.work_on_pending_operations(fibers);
+    }
```

Comment:
Oh, true. Initially, I thought the outer world would be able to send and receive things on-demand and call some kind of `send` or `receive` method directly on the VM, but it probably makes more sense to do everything pull-based (i.e. the outside world can only send stuff into the VM when a fiber listens and otherwise has to take care by itself of queuing up packets to be sent).

Reply:
On second thought: This might work for external channels, but what happens when a channel is sent to the outside?

```candy
main environment =
  channel = core.channel.create 3
  core.channel.send environment.someChannel channel.sendPort
```

In this case, we want to allow the outside world to send stuff to the `channel`. If the channel is full, we still want to allow the outside world to add the pending send to the queue, between other operations performed by fibers.

The difference is that in the queue, some operations need to result in a suspended fiber being woken up and in other cases not – thus, the option.

For sending return values to Candy code, the nursery also uses `send_to_channel` with a performing fiber of `None` because when those sends complete, no fiber should be woken up.